velox/substrait/SubstraitToVeloxPlan.h

```diff
@@ -0,0 +1,134 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "SubstraitToVeloxExpr.h"
+#include "velox/connectors/hive/HiveConnector.h"
+
+namespace facebook::velox::substrait {
+
+/// This class is used to convert the Substrait plan into Velox plan.
+class SubstraitVeloxPlanConverter {
+ public:
+  /// Used to convert Substrait AggregateRel into Velox PlanNode.
+  std::shared_ptr<const core::PlanNode> toVeloxPlan(
+      const ::substrait::AggregateRel& sAgg);
+
+  /// Used to convert Substrait ProjectRel into Velox PlanNode.
+  std::shared_ptr<const core::PlanNode> toVeloxPlan(
+      const ::substrait::ProjectRel& sProject);
+
+  /// Used to convert Substrait FilterRel into Velox PlanNode.
+  std::shared_ptr<const core::PlanNode> toVeloxPlan(
+      const ::substrait::FilterRel& sFilter);
+
+  /// Used to convert Substrait ReadRel into Velox PlanNode.
+  std::shared_ptr<const core::PlanNode> toVeloxPlan(
+      const ::substrait::ReadRel& sRead,
+      u_int32_t& index,
+      std::vector<std::string>& paths,
+      std::vector<u_int64_t>& starts,
+      std::vector<u_int64_t>& lengths);
+
+  /// Used to convert Substrait Rel into Velox PlanNode.
+  std::shared_ptr<const core::PlanNode> toVeloxPlan(
+      const ::substrait::Rel& sRel);
+
+  /// Used to convert Substrait RelRoot into Velox PlanNode.
+  std::shared_ptr<const core::PlanNode> toVeloxPlan(
+      const ::substrait::RelRoot& sRoot);
+
+  /// Used to convert Substrait Plan into Velox PlanNode.
+  std::shared_ptr<const core::PlanNode> toVeloxPlan(
+      const ::substrait::Plan& sPlan);
+
+  /// Will return the index of Partition to be scanned.
+  u_int32_t getPartitionIndex() {
+    return partitionIndex_;
+  }
+
+  /// Will add the paths of the files to be scanned into a vector.
+  void getPaths(std::vector<std::string>& paths) {
+    for (const auto& path : paths_) {
+      paths.emplace_back(path);
+    }
+  }
+
+  /// Will add the starts of the files to be scanned into a vector.
+  void getStarts(std::vector<u_int64_t>& starts) {
+    for (const auto& start : starts_) {
+      starts.emplace_back(start);
+    }
+  }
+
+  /// Will return the lengths to be scanned for each file.
+  void getLengths(std::vector<u_int64_t>& lengths) {
+    for (const auto& length : lengths_) {
+      lengths.emplace_back(length);
+    }
+  }
+
+ private:
+  /// The Partition index.
+  u_int32_t partitionIndex_;
+
+  /// The file paths to be scanned.
+  std::vector<std::string> paths_;
+
+  /// The file starts in the scan.
+  std::vector<u_int64_t> starts_;
+
+  /// The lengths to be scanned.
+  std::vector<u_int64_t> lengths_;
+
+  /// The unique identification for each PlanNode.
+  int planNodeId_ = 0;
+
+  /// The map storing the relations between the function id and the function
+  /// name. Will be constructed based on the Substrait representation.
+  std::unordered_map<uint64_t, std::string> functionMap_;
```

Comment:
This pattern of a function map doesn't seem to correctly represent the variability possible in a Substrait plan. How is this map constructed? It should be done by reading specific other parts of the plan.

Reply:
Yes, I think we need to follow your suggestion to construct the function name into below format:
<function name>:<short_arg_type0>_<short_arg_type1>_..._<short_arg_typeN>
And on your opinion, when should we validate those functions based on yaml specifications? In Spark or in Velox? I think we need to validate them before starting the computing.